src/rosdistro/manifest_provider/git.py

```diff
@@ -48,12 +51,49 @@ def git_manifest_provider(_dist_name, repo, pkg_name):
     assert repo.version
     try:
         release_tag = get_release_tag(repo, pkg_name)
-        package_xml = _get_package_xml(repo.url, release_tag)
-        return package_xml
+        with _temp_git_clone(repo.url, release_tag) as git_repo_path:
+            filename = os.path.join(git_repo_path, 'package.xml')
+            if not os.path.exists(filename):
+                raise RuntimeError('Could not find package.xml in repository "%s"' % repo.url)
+            with open(filename, 'r') as f:
+                return f.read()
     except Exception as e:
         raise RuntimeError('Unable to fetch package.xml: %s' % e)
 
 
+def git_source_manifest_provider(repo):
+    try:
+        with _temp_git_clone(repo.url, repo.version) as git_repo_path:
+            # Include the git hash in our cache dictionary.
+            cmd = [_git_client_executable, 'rev-parse', 'HEAD']
+            result = _run_command(cmd)
+            cache = { '_ref': result['output'] }
+
+            for path, dirnames, filenames in os.walk(git_repo_path):
+                # Don't recurse into hidden folders.
+                if os.path.basename(path)[0] == '.':
+                    dirnames[:] = []
+                    continue
+
+                if 'package.xml' in filenames:
+                    filename = os.path.join(path, 'package.xml')
+                    with open(filename, 'r') as f:
+                        package_xml = f.read()
+
+                    try:
+                        name = parse_package_string(package_xml).name
+                    except InvalidPackage:
+                        raise RuntimeError('Unable to parse package.xml file found in %s' % repo.url)
+                    cache[name] = [ path[len(git_repo_path)+1:], package_xml ]
```

Comment:
This format is different than we talked about. It ends up being:

```
source_repo_xmls:
  my_fancy_repo:
    _ref: b858cb282617fb0956d960215c8e84d1ccf909c6
    pkgname1: ['path/to/pkgname1', '<package>....</package>']
    pkgname2: ['path/to/pkgname1', '<package>....</package>']
    pkgname3: ['path/to/pkgname1', '<package>....</package>']
```

Upon reflection, I didn't feel there was value in trying to support multiple hashes in the cache— we're only ever going to cache the "source" version branch anyway. And if we did have multiple in there in a dictionary, then the consuming code no longer has a "default" option; it _has_ to `git ls-remote` to get a hash in order to resolve, which seems like a big step backward to me.

The source repo cache is only ever going to be best-effort, and this format accepts that reality.